### DISCUSSION TITLE
on failure: don't cleanup, spawn a shell if possible, resume where it left

### DISCUSSION THREAD

**Discussion by hadjiprocopis** [Homebrew/discussions#3429] [Getting started] [answered]
I am just starting using homebrew on OSX 10.13 (unsupported).
How can I tell homebrew that when it encounters an error compiling some package (case in point: llvm) to leave its build-dir just there (and not to cleanup) so that I can cd there and see if I can do something to fix the issue. Then I would like to tell homebrew to somehow resume compilation (or restart compilation) after my fix in that same folder.

For example, it would be ideal if there is a way that homebrew spawns a shell into the build-dir, where I can see if  I can fix, then on exiting the shell, homebrew resumes or just re-configures and re-compiles without a make-clean (I can do that myself if need be via that shell).

An alternative would be that homebrew  does not cleanup the build-dir of a failed compilation, and it accepts specifying an already existing build-dir where it should resume compilation.

ps. build-dir=package's build dir

**Comment by carlocab** [comment, marked as answer]
You might be interested in these flags. See `brew help install` or `man brew` for details.

```
       -d, --debug
              If brewing fails, open an interactive debugging session with access to IRB or a shell inside the temporary build
              directory.
...
       --keep-tmp
              Retain the temporary files created during installation.
...
       -i, --interactive
              Download and patch formula, then open a shell. This allows the user to run ./configure --help and otherwise determine how
              to turn the software package into a Homebrew package.

```

**Reply by hadjiprocopis** [reply]
perfect thank you!

**Comment by hadjiprocopis** [comment]
I need a bit more information on ```--debug``` and ```--interactive```.

Firstly, where can I see what the options in the menu do exactly? (e.g. raise/ignore/irb etc.).

Secondly, after fixing the compilation issues, I need to give back control to homebrew for doing the installation, the linking etc. Let's say I run brew with ```--debug```. At some point compilation fails and I am presented with the menu. I ask for ```shell```. I am now in the build-dir, I edit some files, and compilation can now be resumed. I type ```exit``` to exit the shell brew gave me. And I am back to said menu. Now I need brew to resume compilation from where it left it. How?

Another scenario: I need brew to restart the compilation WITHOUT erasing the previous build-dir where my patches are applied but in the same build-dir I modified files in. Alternatively, I can resume the compilation myself with a ```make``` (caveat: what if the make command brew used has extra options I am not aware of?). Let's say ```make``` compiles OK and once all is correct I press exit on the shell. Now back to shell, which option to use?

If there are any information on how to achieve the above  please point me to it or I would appreciate some tips when you have the time.

Thank you.

**Reply by SMillerDev** [reply]
> Now I need brew to resume compilation from where it left it. How?

This is not possible, `--debug` is there to debug the issues. Any fixes will need to be done in the brew formula afterwards to be applied.

**Reply by hadjiprocopis** [reply]
Would the following "workaround" work:?
After a failure, the ```--debug``` menu is presented. I don't press anything. On another terminal I change to the build-dir and apply fixes. I can even issue a ```make``` to see if things worked as expected. If all works, then on the debug menu I press ```ignore```. I guess then it will restart compilation?

p.s. I am on unsupported version 10.13 and I am trying hard to adopt homebrew after a long time with macports (which seems to have failed me because of the legacy OS). I flatly refuse to throw away perfectly good hardware, so excuse me and thank you for bearing with me. I have just managed to install latest LLVM and now installing latest GCC.

**Reply by SMillerDev** [reply]
`--debug` will never continue compiling AFAIK since it's only used to show you where the problem is. Not resolve it.

**Comment by carlocab** [comment]
Homebrew isn't designed for what you're trying to do, so there isn't a straightforward way to do what you're asking for.

One thing that might help you do part of what you're looking for is to use the `--keep-tmp` flag. `brew` will tell you the location of the temporary build directory when you do this and it exits from a formula build (successfully or not).

In the event of failure, you can modify the relevant formula as needed and then add
```ruby
cd "temporary-directory-used-from-previous-build-attempt"
```
to the appropriate place inside the `install` method of the formula. This gives you a somewhat limited way to "resume" a failed build from where you left it.

That said, what you want to do might be best done without using a package manager -- I'm not aware of any (whether on macOS or Linux) that has the features you're looking for here.

**Comment by hadjiprocopis** [comment]
A few years later, on different formulae but still the same OSX 10.13 I had problems again with failed installs.

I am going to outline how I managed to install them. The first problem was with icu4c@77. File `i18n/measunit_extra.cpp` fails compilation because there is no `abs(double)` in the included headers. But, for example there is `fabs` in `math.h`. The second problem is with `gcc15.1.0` which at first failed because of building documentation failed, fixed by adding `--disable-docs` in the `configure` step. When that option is added, it compiles when built manually but fails when I `brew edit gcc`, adding the configuration option, save and then `HOMEBREW_NO_INSTALL_FROM_API=1 brew update --verbose gcc`. I don't remember where `gmake` failed, each time differently, so I suggest `HOMEBREW_MAKE_JOBS=1 brew update ...`

So this is the trick I did, obviously you may frown upon it:

`brew edit gcc` : will open the formula in an editor, under `def install` amend the configuration parameters (by adding e.g. `--disable-docs`) and then, the crucial step is to find where the configure step is (for example under `mkdir "build" do`) add a line after the configure step, like this: 
   ```
system "../configure", *args
# add these 3 lines:
echo "open another shell, and do gmake with ",  *make_args
echo "press a key in this shell when your gmake is done..."
system "/usr/bin/read"
# comment out this line
#system "gmake", *make_args
```
edit: how can I print the current tmp dir brew is using for the install so that the user can go to it?

So, essentially run the configure via `brew` and then pause. In another shell run `gmake` fixing problems as you find them and when this is all done, go back to the first shell and press a key in order to let `brew` continue with the installation into the catacombs of the homebrew. I think this is the closest to what I have asked initially, i.e. when a compilation error is encountered pause `brew` and give me a shell in order to fix the (most likely silly) issue.

I am happy that i not only give life to my OSX 10.13 but I can use it as my dev machine when travelling
